File: incoming/decompressed/evo-tactics/tools/py/validate_species.py
#!/usr/bin/env python3
import sys, json, yaml, argparse
from collections import defaultdict

def load_yaml(path):
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)

def collect_catalog(spec):
    slots = spec["catalog"]["slots"]
    by_slot = {slot: set(parts.keys()) for slot, parts in slots.items()}
    synmap = {s["id"]: s for s in spec.get("catalog", {}).get("synergies", [])}
    return by_slot, synmap

def get_caps(spec):
    caps = spec.get("global_rules", {}).get("stacking_caps", {})
    res_cap = caps.get("res_cap_per_type", 2)
    dr_cap = caps.get("dr_cap_per_type", 1)
    return res_cap, dr_cap

def get_globals(spec):
    gr = spec.get("global_rules", {})
    return {
        "biomes": set(gr.get("biomes", [])),
        "affixes": set(gr.get("biome_affixes", [])),
        "roles": set(gr.get("spawn_roles", [])),
        "vc_indices": set(gr.get("vc_indices", [])),
        "mating_keys": ["env","structure","security","resources","privacy"],
    }

def gather_chosen_parts(dp):
    chosen = set()
    for slot in ["locomotion","metabolism"]:
        if dp.get(slot):
            chosen.add(f"{slot}.{dp[slot]}")
    for slot in ["offense","defense","senses"]:
        for pid in dp.get(slot, []) or []:
            chosen.add(f"{slot}.{pid}")
    return chosen

def resolve_part(spec, slot, pid):
    return spec["catalog"]["slots"][slot].get(pid)

def accumulate_resistances(spec, dp):
    res_sum, vuln_sum, dr_sum = defaultdict(int), defaultdict(int), defaultdict(int)
    def acc(slot, pid):
        part = resolve_part(spec, slot, pid)
        if not part: return
        eff = part.get("effects", {})
        r = eff.get("resistances", {})
        for t,v in (r.get("res") or {}).items():   res_sum[t]  += int(v)
        for t,v in (r.get("vuln") or {}).items():  vuln_sum[t] += int(v)
        for t,v in (r.get("dr") or {}).items():    dr_sum[t]   += int(v)
    for s in ["locomotion","metabolism"]:
        if dp.get(s): acc(s, dp[s])
    for s in ["offense","defense","senses"]:
        for pid in dp.get(s, []) or []: acc(s, pid)
    return dict(res_sum), dict(vuln_sum), dict(dr_sum)

def compute_active_synergies(spec, chosen_set):
    active = []
    for syn in spec.get("catalog",{}).get("synergies",[]):
        reqs = set(syn.get("when_all", []))
        if all(req in chosen_set for req in reqs):
            active.append(syn["id"])
    return active

def compute_known_counters(spec, chosen_set):
    out = []
    for item in spec.get("global_rules",{}).get("counters_reference", []):
        ctr_parts = item.get("counters", [])
        hit = any(any(k.endswith("."+c) or c in k for k in chosen_set) for c in ctr_parts)
        if hit: out.append(item["counter"])
    return out

def validate_biomes_and_affixes(globals_, sp, warnings, errors):
    unknown_biomes = [b for b in sp.get("biomes", []) if b not in globals_["biomes"]]
    if unknown_biomes:
        errors.append(f"unknown biomes: {unknown_biomes}")
    bm = sp.get("biome_mods", {})
    for k in ("mood","diff_base_mod"):
        for b in (bm.get(k, {}) or {}).keys():
            if b not in globals_["biomes"]:
                errors.append(f"biome_mods.{k}: unknown biome '{b}'")
    for a in (bm.get("affix_bias", {}) or {}).keys():
        if a not in globals_["affixes"]:
            errors.append(f"biome_mods.affix_bias: unknown affix '{a}'")

def validate_spawn_roles(globals_, sp, warnings, errors):
    rp = sp.get("spawn_profile", {})
    rw = rp.get("role_weights", {}) or {}
    bad = [r for r in rw.keys() if r not in globals_["roles"]]
    if bad:
        errors.append(f"spawn_profile.role_weights: unknown roles {bad}")
    if any((not isinstance(v, int) or v < 0) for v in rw.values()):
        errors.append("spawn_profile.role_weights: values must be non-negative integers")

def validate_vc_signature(globals_, sp, warnings, errors):
    vc = sp.get("vc_signature")
    if not vc:
        warnings.append("vc_signature missing")
        return
    extra = set(vc.keys()) - globals_["vc_indices"]
    missing = globals_["vc_indices"] - set(vc.keys())
    if extra:
        errors.append(f"vc_signature: unknown indices {sorted(extra)}")
    if missing:
        warnings.append(f"vc_signature: missing indices {sorted(missing)}")
    if vc:
        for k,v in vc.items():
            try:
                f = float(v)
                if f < 0 or f > 1: warnings.append(f"vc_signature.{k} out of range [0,1]: {v}")
            except Exception:
                errors.append(f"vc_signature.{k} not numeric: {v}")

def validate_nest_profile(sp, warnings, errors):
    np = sp.get("nest_profile")
    if not np:
        warnings.append("nest_profile missing")
        return
    required = ["env","structure","security","resources","privacy"]
    missing = [k for k in required if k not in np]
    if missing:
        errors.append(f"nest_profile: missing keys {missing}")
        return
    env = np.get("env", {})
    if not isinstance(env, dict) or any(k not in env for k in ["temp","humidity","light"]):
        warnings.append("nest_profile.env should include temp/humidity/light")
    if not isinstance(np.get("structure"), list) or not all(isinstance(x,str) for x in np["structure"]):
        warnings.append("nest_profile.structure should be a list of strings")
    if not isinstance(np.get("security"), int):
        warnings.append("nest_profile.security should be an integer")
    if not isinstance(np.get("resources"), list) or not all(isinstance(x,str) for x in np["resources"]):
        warnings.append("nest_profile.resources should be a list of strings")
    if not isinstance(np.get("privacy"), bool):
        warnings.append("nest_profile.privacy should be boolean")

def validate(path):
    spec = load_yaml(path)
    by_slot, synmap = collect_catalog(spec)
    res_cap, dr_cap = get_caps(spec)
    globals_ = get_globals(spec)

    report = {"file": path, "species": [], "errors": 0, "warnings": 0}
    for sp in spec.get("species", []):
        sid = sp["id"]; dp = sp.get("default_parts", {})
        errors, warnings = [], []

        est = sp.get("estimated_weight", 0)
        bud = sp.get("weight_budget", spec["global_rules"]["morph_budget"]["default_weight_budget"])
        over_budget = est > bud
        if over_budget: warnings.append(f"estimated_weight {est} exceeds budget {bud}")

        for slot in ["locomotion","metabolism"]:
            pid = dp.get(slot)
            if pid and pid not in by_slot.get(slot, set()):
                errors.append(f"missing part: {slot}.{pid}")
        for slot in ["offense","defense","senses"]:
            for pid in dp.get(slot, []) or []:
                if pid not in by_slot.get(slot, set()):
                    errors.append(f"missing part: {slot}.{pid}")

        for syn in sp.get("synergy_hints", []) or []:
            if syn not in synmap:
                warnings.append(f"synergy hint '{syn}' not in catalog.synergies")

        res_sum, vuln_sum, dr_sum = accumulate_resistances(spec, dp)
        for t,v in res_sum.items():
            if v > res_cap: warnings.append(f"res cap exceeded: {t}={v} > {res_cap}")
        for t,v in dr_sum.items():
            if v > dr_cap: warnings.append(f"dr cap exceeded: {t}={v} > {dr_cap}")

        validate_biomes_and_affixes(globals_, sp, warnings, errors)
        validate_spawn_roles(globals_, sp, warnings, errors)
        validate_vc_signature(globals_, sp, warnings, errors)
        validate_nest_profile(sp, warnings, errors)

        chosen_set     = gather_chosen_parts(dp)
        active_syn     = compute_active_synergies(spec, chosen_set)
        known_counters = compute_known_counters(spec, chosen_set)

        report["species"].append({
            "id": sid,
            "display_name": sp.get("display_name", sid),
            "budget": {"used": est, "max": bud, "over_budget": over_budget},
            "parts_ok": len([e for e in errors if e.startswith("missing part")]) == 0,
            "active_synergies": active_syn,
            "synergy_hints": sp.get("synergy_hints", []),
            "known_counters": known_counters,
            "res_summary": res_sum,
            "vuln_summary": vuln_sum,
            "dr_summary": dr_sum,
            "warnings": warnings,
            "errors": errors
        })
        report["errors"]   += len(errors)
        report["warnings"] += len(warnings)

    report["ui_summary"] = [{
        "id": s["id"],
        "name": s["display_name"],
        "budget": f'{s["budget"]["used"]}/{s["budget"]["max"]}',
        "over_budget": s["budget"]["over_budget"],
        "active_synergies": s["active_synergies"],
        "known_counters": s["known_counters"],
        "warnings": s["warnings"]
    } for s in report["species"]]

    import json
    print(json.dumps(report, ensure_ascii=False, indent=2))
    return 0 if report["errors"] == 0 else 1

if __name__ == "__main__":
    import argparse
    ap = argparse.ArgumentParser()
    ap.add_argument("path", nargs="?", default="data/species.yaml")
    args = ap.parse_args()
    import sys
    sys.exit(validate(args.path))

File: incoming/decompressed/evo-tactics/tools/ts/validate_species.ts
#!/usr/bin/env node
import fs from "fs";
import yaml from "js-yaml";

function loadYaml(path) { return yaml.load(fs.readFileSync(path, "utf8")); }
function collectCatalog(spec) {
  const slots = spec.catalog.slots;
  const bySlot = {};
  Object.keys(slots).forEach(s => bySlot[s] = new Set(Object.keys(slots[s])));
  const synMap = {};
  (spec.catalog.synergies || []).forEach(s => synMap[s.id] = s);
  return { bySlot, synMap };
}
function getCaps(spec) {
  const caps = spec.global_rules?.stacking_caps || {};
  return { resCap: caps.res_cap_per_type ?? 2, drCap: caps.dr_cap_per_type ?? 1 };
}
function getGlobals(spec) {
  const gr = spec.global_rules || {};
  return {
    biomes: new Set(gr.biomes || []),
    affixes: new Set(gr.biome_affixes || []),
    roles: new Set(gr.spawn_roles || []),
    vc_indices: new Set(gr.vc_indices || [])
  };
}
function gatherChosen(dp) {
  const out = new Set();
  ["locomotion","metabolism"].forEach(sl => { if (dp[sl]) out.add(`${sl}.${dp[sl]}`); });
  ["offense","defense","senses"].forEach(sl => (dp[sl]||[]).forEach(p=> out.add(`${sl}.${p}`)));
  return out;
}
function resolvePart(spec, slot, pid) { return spec.catalog.slots[slot]?.[pid] ?? null; }
function accumulateRes(spec, dp) {
  const res={}, vuln={}, dr={};
  const acc = (slot,pid)=>{
    const part = resolvePart(spec, slot, pid); if (!part) return;
    const eff = part.effects || {}; const ro = eff.resistances || {};
    Object.entries(ro.res   || {}).forEach(([t,v])=> res[t]=(res[t]||0)+Number(v));
    Object.entries(ro.vuln  || {}).forEach(([t,v])=> vuln[t]=(vuln[t]||0)+Number(v));
    Object.entries(ro.dr    || {}).forEach(([t,v])=> dr[t]=(dr[t]||0)+Number(v));
  };
  ["locomotion","metabolism"].forEach(s => { if (dp[s]) acc(s, dp[s]); });
  ["offense","defense","senses"].forEach(s => (dp[s]||[]).forEach(p=> acc(s,p)));
  return {res, vuln, dr};
}
function computeActiveSynergies(spec, chosen) {
  const out = [];
  (spec.catalog.synergies || []).forEach(syn=>{
    const reqs = syn.when_all || [];
    if (reqs.every(r => chosen.has(r))) out.push(syn.id);
  });
  return out;
}
function computeKnownCounters(spec, chosen) {
  const out = [];
  (spec.global_rules?.counters_reference || []).forEach(item=>{
    const hit = (item.counters || []).some(c=>{
      for (const k of Array.from(chosen)) if (k.endsWith("."+c) || k.includes(c)) return true;
      return false;
    });
    if (hit) out.push(item.counter);
  });
  return out;
}

function validateBiomesAndAffixes(globals, sp, warnings, errors) {
  const biomes = sp.biomes || [];
  const badBiomes = biomes.filter(b => !globals.biomes.has(b));
  if (badBiomes.length) errors.push(`unknown biomes: ${JSON.stringify(badBiomes)}`);
  const bm = sp.biome_mods || {};
  ["mood","diff_base_mod"].forEach(k => {
    Object.keys(bm[k] || {}).forEach(b => {
      if (!globals.biomes.has(b)) errors.push(`biome_mods.${k}: unknown biome '${b}'`);
    });
  });
  Object.keys(bm.affix_bias || {}).forEach(a => {
    if (!globals.affixes.has(a)) errors.push(`biome_mods.affix_bias: unknown affix '${a}'`);
  });
}
function validateSpawnRoles(globals, sp, warnings, errors) {
  const rw = (sp.spawn_profile && sp.spawn_profile.role_weights) || {};
  const bad = Object.keys(rw).filter(r => !globals.roles.has(r));
  if (bad.length) errors.push(`spawn_profile.role_weights: unknown roles ${JSON.stringify(bad)}`);
  if (Object.values(rw).some(v=> typeof v !== "number" || v < 0)) {
    errors.push("spawn_profile.role_weights: values must be non-negative integers");
  }
}
function validateVcSignature(globals, sp, warnings, errors) {
  const vc = sp.vc_signature;
  if (!vc) { warnings.push("vc_signature missing"); return; }
  const keys = new Set(Object.keys(vc));
  const extra = [...keys].filter(k => !globals.vc_indices.has(k));
  const missing = [...globals.vc_indices].filter(k => !keys.has(k));
  if (extra.length) errors.push(`vc_signature: unknown indices ${JSON.stringify(extra)}`);
  if (missing.length) warnings.push(`vc_signature: missing indices ${JSON.stringify(missing)}`);
  Object.entries(vc).forEach(([k,v])=>{
    const f = Number(v);
    if (Number.isNaN(f)) errors.push(`vc_signature.${k} not numeric: ${v}`);
    else if (f<0 || f>1) warnings.push(`vc_signature.${k} out of range [0,1]: ${v}`);
  });
}
function validateNestProfile(sp, warnings, errors) {
  const np = sp.nest_profile;
  if (!np) { warnings.push("nest_profile missing"); return; }
  const req = ["env","structure","security","resources","privacy"];
  const missing = req.filter(k => !(k in np));
  if (missing.length) { errors.push(`nest_profile: missing keys ${JSON.stringify(missing)}`); return; }
  if (typeof np.env !== "object" || !("temp" in np.env) || !("humidity" in np.env) || !("light" in np.env)) {
    warnings.push("nest_profile.env should include temp/humidity/light");
  }
  if (!Array.isArray(np.structure) || np.structure.some(x=> typeof x !== "string")) {
    warnings.push("nest_profile.structure should be a list of strings");
  }
  if (typeof np.security !== "number") warnings.push("nest_profile.security should be an integer");
  if (!Array.isArray(np.resources) || np.resources.some(x=> typeof x !== "string")) {
    warnings.push("nest_profile.resources should be a list of strings");
  }
  if (typeof np.privacy !== "boolean") warnings.push("nest_profile.privacy should be boolean");
}

function validate(path) {
  const spec = loadYaml(path);
  const { bySlot, synMap } = collectCatalog(spec);
  const caps = spec.global_rules?.stacking_caps || {};
  const resCap = caps.res_cap_per_type ?? 2;
  const drCap  = caps.dr_cap_per_type  ?? 1;
  const globals = getGlobals(spec);

  const report = { file: path, species: [], errors: 0, warnings: 0 };

  (spec.species || []).forEach(sp=>{
    const sid = sp.id, dp = sp.default_parts || {};
    const errors = [], warnings = [];

    const est = sp.estimated_weight ?? 0;
    const bud = sp.weight_budget ?? spec.global_rules.morph_budget.default_weight_budget;
    const over_budget = est > bud;
    if (over_budget) warnings.push(`estimated_weight ${est} exceeds budget ${bud}`);

    ["locomotion","metabolism"].forEach(sl=>{
      const pid = dp[sl];
      if (pid && !bySlot[sl]?.has(pid)) errors.push(`missing part: ${sl}.${pid}`);
    });
    ["offense","defense","senses"].forEach(sl=>{
      (dp[sl]||[]).forEach(pid=>{
        if (!bySlot[sl]?.has(pid)) errors.push(`missing part: ${sl}.${pid}`);
      });
    });

    (sp.synergy_hints||[]).forEach(syn=> {
      if (!synMap[syn]) warnings.push(`synergy hint '${syn}' not in catalog.synergies`);
    });

    // resistances
    const res={}, vuln={}, dr={};
    const acc=(slot,pid)=>{
      const part = spec.catalog.slots[slot]?.[pid]; if (!part) return;
      const eff = part.effects || {}; const ro = eff.resistances || {};
      Object.entries(ro.res   || {}).forEach(([t,v])=> res[t]=(res[t]||0)+Number(v));
      Object.entries(ro.vuln  || {}).forEach(([t,v])=> vuln[t]=(vuln[t]||0)+Number(v));
      Object.entries(ro.dr    || {}).forEach(([t,v])=> dr[t]=(dr[t]||0)+Number(v));
    };
    ["locomotion","metabolism"].forEach(s=> dp[s] && acc(s, dp[s]));
    ["offense","defense","senses"].forEach(s=> (dp[s]||[]).forEach(pid=> acc(s,pid)));

    Object.entries(res).forEach(([t,v])=> { if (v > resCap) warnings.push(`res cap exceeded: ${t}=${v} > ${resCap}`); });
    Object.entries(dr).forEach(([t,v])=>  { if (v > drCap)  warnings.push(`dr cap exceeded: ${t}=${v} > ${drCap}`); });

    validateBiomesAndAffixes(globals, sp, warnings, errors);
    validateSpawnRoles(globals, sp, warnings, errors);
    validateVcSignature(globals, sp, warnings, errors);
    validateNestProfile(sp, warnings, errors);

    const chosen = gatherChosen(dp);
    const activeSyn     = computeActiveSynergies(spec, chosen);
    const knownCounters = computeKnownCounters(spec, chosen);

    report.species.push({
      id: sid,
      display_name: sp.display_name || sid,
      budget: { used: est, max: bud, over_budget },
      parts_ok: errors.filter(e=>e.startsWith("missing part")).length===0,
      active_synergies: activeSyn,
      synergy_hints: sp.synergy_hints || [],
      known_counters: knownCounters,
      res_summary: res,
      vuln_summary: vuln,
      dr_summary: dr,
      warnings, errors
    });
    report.errors   += errors.length;
    report.warnings += warnings.length;
  });

  report.ui_summary = report.species.map(s=> ({
    id: s.id,
    name: s.display_name,
    budget: `${s.budget.used}/${s.budget.max}`,
    over_budget: s.budget.over_budget,
    active_synergies: s.active_synergies,
    known_counters: s.known_counters,
    warnings: s.warnings
  }));

  console.log(JSON.stringify(report, null, 2));
  process.exit(report.errors === 0 ? 0 : 1);
}

const pathArg = process.argv[2] || "data/species.yaml";
validate(pathArg);
